[GEN][ZH] Reset DrawableID on map load to avoid massive id growth over the lifetime of the game

Mauller opened the pull request:
This PR resets the drawable ID within `GameClient::reset()` so it does not keep growing over the lifetime of the game. The implication of uncapped growth is that the `GameClient::m_drawableVector` will keep growing as well.

EDIT: This PR has been heavily modified over time and now only resets the DrawableID

~~Within GameClient::reset the lookup tables were being cleared before their relevant drawables were removed from them.
This was causing out of range container errors similar to what was seen in GameLogic::reset.~~

~~destroyDrawable was also reordered so that the drawable was removed from the lookup table before being destroyed in the drawable list.~~

~~This PR reverts some earlier changes made when zero hour was using a vector as a lookup table and resets the drawable ID in gameclient.~~

~~It also adds some debug messages to GameClient and GameLogics ```removeObject/DrawableFromLookupTable``` functions.~~

Mauller commented:
Just for some extra context, if you leave the game on the shell map for at least two loops then start a skirmish, you will hit the problem.

xezon commented on Generals/Code/GameEngine/Source/GameClient/GameClient.cpp:
```diff
@@ -430,12 +430,6 @@ void GameClient::init( void )
-#if USING_STLPORT
-	m_drawableHash.resize(DRAWABLE_HASH_SIZE);
-#else
-	m_drawableHash.reserve(DRAWABLE_HASH_SIZE);
```
On move, reserve was silently changed to rehash. Why?

https://en.cppreference.com/w/cpp/container/unordered_map/rehash
https://en.cppreference.com/w/cpp/container/unordered_map/reserve

vs2022 xhash for unordered_map:
```cpp
    void rehash(size_type _Buckets) { // rebuild table with at least _Buckets buckets
        // don't violate a.bucket_count() >= a.size() / a.max_load_factor() invariant:
        _Buckets = (_STD max)(_Min_load_factor_buckets(_List.size()), _Buckets);
        if (_Buckets <= _Maxidx) { // we already have enough buckets; nothing to do
            return;
        }

        _Forced_rehash(_Buckets);
    }

    void reserve(size_type _Maxcount) { // rebuild table with room for _Maxcount elements
        rehash(_Min_load_factor_buckets(_Maxcount));
    }
```

stlport hashtable for hash_map:

```cpp
template <class _Val, class _Key, class _HF, class _ExK, class _EqK, class _All>
void hashtable<_Val,_Key,_HF,_ExK,_EqK,_All>
  ::resize(size_type __num_elements_hint)
{
  const size_type __old_n = _M_buckets.size();
  if (__num_elements_hint > __old_n) {
    const size_type __n = _M_next_size(__num_elements_hint);
    if (__n > __old_n) {
      _BucketVector __tmp(__n, (void*)(0),
			  _M_buckets.get_allocator());
      _STLP_TRY {
        for (size_type __bucket = 0; __bucket < __old_n; ++__bucket) {
          _Node* __first = (_Node*)_M_buckets[__bucket];
          while (__first) {
            size_type __new_bucket = _M_bkt_num(__first->_M_val, __n);
            _M_buckets[__bucket] = __first->_M_next;
            __first->_M_next = (_Node*)__tmp[__new_bucket];
            __tmp[__new_bucket] = __first;
            __first = (_Node*)_M_buckets[__bucket];          
          }
        }
        _M_buckets.swap(__tmp);
      }
#         ifdef _STLP_USE_EXCEPTIONS
      catch(...) {
        for (size_type __bucket = 0; __bucket < __tmp.size(); ++__bucket) {
          while (__tmp[__bucket]) {
            _Node* __next = ((_Node*)__tmp[__bucket])->_M_next;
            _M_delete_node((_Node*)__tmp[__bucket]);
            __tmp[__bucket] = __next;
          }
        }
        throw;
      }
#         endif /* _STLP_USE_EXCEPTIONS */
    }
  }
}
```

Which one is closer to what this aims to do? I assumed `reserve` is closer to the desired state.

Mauller replied:
This one was actually a mistake but would not have affected things in the long run.
I was experimenting between using reserve and rehash as some sources claimed rehash could also free underlying memory.

Reverting it back to resize.

xezon commented on Generals/Code/GameEngine/Source/GameClient/GameClient.cpp:
```diff
@@ -457,6 +459,9 @@ void GameClient::reset( void )
 	// clear any drawable TOC we might have
 	m_drawableTOC.clear();
 
+	// TheSuperHackers @fix Mauller nextDrawableID was never being reset.
```
`// TheSuperHackers @fix Mauller 13/04/2025 Resets drawable id so it does not keep growing over the lifetime of a game session.`

xezon commented on Generals/Code/GameEngine/Source/GameClient/GameClient.cpp:
```diff
@@ -448,6 +442,14 @@ void GameClient::reset( void )
 	}
 	m_drawableList = NULL;
 
+	// TheSuperHackers @fix Mauller 12/04/25 Drawables need to be destroyed before clearing the drawable lookup table.
```
Although this is the code that was moved, the comment is misplaced, because the drawable destroy does not happen here, but a few lines above. This comment is better placed below `// destroy all Drawables`

If the comment is not meant to be moved, then it needs to be rephrased.

Mauller replied:
Rephrased the comment so it makes more sense with the ordering that needs to occur.

xezon commented on Generals/Code/GameEngine/Source/GameClient/GameClient.cpp:
```diff
@@ -448,6 +442,14 @@ void GameClient::reset( void )
 	}
 	m_drawableList = NULL;
 
+	// TheSuperHackers @fix Mauller 12/04/25 Drawables need to be destroyed before clearing the drawable lookup table.
```
12/04/2025

xezon commented on Generals/Code/GameEngine/Source/GameClient/GameClient.cpp:
```diff
@@ -785,6 +790,10 @@ void GameClient::destroyDrawable( Drawable *draw )
 	TheInGameUI->disregardDrawable( draw );
 
+	// remove the drawable from our hash of drawables
+	// TheSuperHackers @fix Mauller 12/04/2025 Need to remove drawable from lookup table before destroying it in the main list.
```
This move does not look necessary. `draw` is not destroyed before `draw->deleteInstance();`. What was the observed runtime error?

Mauller replied:
Reverting this, the issue was back in ::Reset when the drawable lookup table was cleared before the drawable was being erased. 

This was an earlier change that i though was helping, but having looked back over the code it does not make a differenc since the problem was elsewhere.

xezon commented on GeneralsMD/Code/GameEngine/Source/GameClient/GameClient.cpp:
```diff
@@ -467,6 +462,12 @@ void GameClient::reset( void )
 	}
 	m_drawableList = NULL;
 
+	// TheSuperHackers @fix Mauller 12/04/25 Drawables need to be destroyed before clearing the drawable lookup table.
```
I am very suspicious of this change. What was the observed problem?

`m_drawableVector` simply contains pointers to drawables. It looks reasonable to purge these before deleting them, otherwise the pointers dangle between deletion and `m_drawableVector` purge.

I suspect the actual issue is in
```cpp
void GameClient::removeDrawableFromLookupTable( Drawable *draw )
{
	// sanity
	if( draw == NULL )
		return;

	// remove from table
//	m_drawableHash.erase( draw->getID() );
	m_drawableVector[ draw->getID() ] = NULL;
}
```

Notice how this wants to write to the lookup table even if there was no record. It needs checking:

```cpp
void GameClient::removeDrawableFromLookupTable( Drawable *draw )
{
	// sanity
	if( draw == NULL )
		return;

    if( draw->getID() < m_drawableVector.size() )
        m_drawableVector[ draw->getID() ] = NULL;
}
```

The same principle applies to the hash table version in Generals.

Mauller replied:
The problem is that by clearing the drawbles table, during the loop of destroying drawables, destroyDrawable( draw ) gets called which internally tries to null the drawable in the lookup table.

So like your change with GameLogic::Reset, it is cleaner to just destroy the objects first then reset the lookup table.

xezon replied:
Yes, and that is why I suggest to fix the actual offender in function `removeDrawableFromLookupTable`, and do not reorder the cleanup, because the cleanup order is correct.

Mauller replied:
This is only a problem when the list gets cleared, you would be constantly hitting the conditional check when under normal runtime where this is not an issue.

This is no different to the fix you made here https://github.com/TheSuperHackers/GeneralsGameCode/pull/645/commits/bf19942d7296ad21cafeea5daaa0dfebf270a567 when it was the exact same situation.

xezon replied:
Ok. Then please put an assert in `removeDrawableFromLookupTable` to protect against misuse.

Mauller replied:
Updated with debug assert with a message showing the bad object ID and current size of the lookup table.

Mauller commented:
Updated with corrections.

Mauller commented:
Updated with Debug Assert check on drawable ID in removeDrawableFromLookupTable.

xezon commented on GeneralsMD/Code/GameEngine/Source/GameClient/GameClient.cpp:
```diff
@@ -877,6 +881,8 @@ void GameClient::addDrawableToLookupTable(Drawable *draw )
 void GameClient::removeDrawableFromLookupTable( Drawable *draw )
 {
 
+	DEBUG_ASSERTCRASH( draw->getID() < m_drawableVector.size(), ("bad drawableID: %u when removing drawable from lookup table of size: %u", (UnsignedInt)draw->getID(), (UnsignedInt)m_drawableVector.size()) );
```
This needs to be placed behind draw null check.

Mauller replied:
i will change it after the reversion to using a hashmap in zero hour as i will need to alter how the assert works.

Mauller commented:
Okay, hopefully one last push, i didn't realise that STL Ports hashmap doesn't have a contains function, so had to wrap it in a conditional so only non stlport builds use the debug statement

xezon commented on Generals/Code/GameEngine/Source/GameClient/GameClient.cpp:
```diff
@@ -838,6 +843,10 @@ void GameClient::removeDrawableFromLookupTable( Drawable *draw )
 		return;
 
+#ifndef USING_STLPORT
+	DEBUG_ASSERTCRASH( m_drawableHash.contains(draw->getID()) , ("bad DrawableID: %u when removing drawable from lookup table, object not found.", (UnsignedInt)draw->getID()) );
```
`m_drawableHash.find(id) != m_drawableHash.end()`, then `USING_STLPORT` can be removed.

Mauller replied:
Can change that later.

xezon commented on Generals/Code/GameEngine/Source/GameLogic/System/GameLogic.cpp:
```diff
@@ -3335,6 +3335,10 @@ void GameLogic::removeObjectFromLookupTable( Object *obj )
 		return;
 
+#ifndef USING_STLPORT
+	DEBUG_ASSERTCRASH( m_objHash.contains(obj->getID()) , ("bad ObjectID: %u when removing object from lookup table, object not found.", (UnsignedInt)obj->getID()) );
```
Underlying type of id's are signed int, so can stay with that.

Questionable choice that they are enums instead of UnsignedInt typedefs to begin with, but that is another story.

Mauller replied:
i think they are only signed int due to them being based on an enum where they define 0 as an invalid object ID and 7F as max (for obvious reasons).

xezon replied:
%d
(Int)obj->getID()

xezon commented on GeneralsMD/Code/GameEngine/Source/GameClient/GameClient.cpp:
```diff
@@ -468,6 +462,14 @@ void GameClient::reset( void )
 	}
 	m_drawableList = NULL;
 
+	// TheSuperHackers @fix Mauller 12/04/2025 Resetting the drawables lookup table needs to occur after destroying the drawables.
```
With the hash map, is this really a fix now? Was the problem not exclusive to the vector version?

xezon replied:
Yes reverting the order will require the removal of the assert again. At least from that particular function.

Mauller replied:
The assert doesn't really make sense beyond this function, so might be a case to just reveting the clear order changes in this PR and only keep the changes that reset the drawable ID.

Mauller replied:
seems to work fine still where it is.

xezon commented on GeneralsMD/Code/GameEngine/Source/GameClient/GameClient.cpp:
```diff
@@ -468,6 +462,14 @@ void GameClient::reset( void )
 	}
 	m_drawableList = NULL;
 
+	// TheSuperHackers @fix Mauller 12/04/2025 Resetting the drawables lookup table needs to occur after destroying the drawables.
```
I do wonder if this move is actually a minor performance degradation now. Because `destroyDrawable` will need to erase each element from the `m_drawableHash` one by one, whereas `m_drawableHash.clear()` does it in one swoop.

That said, perhaps #645 can be reverted as well now. Can you test?

Mauller replied:
I think inevitably it won't make a significant difference.

The access to check if the map contains data related to the requested key will still occur when the removal from the lookup table occurs inside the destroy object/drawable loops.

But if we revert the ordering change then the debug output will get constantly hit during a reset as the table has been cleared already.

Mauller replied:
Actually, thinking about it a bit more, if the clear order gets reverted so the hashmap is cleared before destroying objects, the assert will always get triggered during reset on every object that is destroyed no matter where the assert is placed.

Mauller commented:
Reverted the change made by #645 as it was no longer required after #676.

Also updated based on reccomendations. Should be good now and works fine.

Mauller commented:
Rebased with main after recent reversions and fixes to the vector based lookup tables in zero hour.

I have removed the asserts in the end as these are essentially redundant now.

This PR and commit only resets the drawable ID so should be a clean merge.

xezon reviewed (approved):
Looks good. Does this cause no mismatch?

Mauller commented:
> Looks good. Does this cause no mismatch?

If this caused mismatches it would be doing so in vanilla all of the time as no ones DrawableID would be synchronised.
